### PR TITLE
Add Dan Cross's slice functions to libc.

### PR DESCRIPTION
**Pull request by rminnich**
They're required for devacpi.

### CODE REVIEW

File: sys/include/libc.h [hunk 1]
```diff
@@ -758,3 +758,34 @@ extern char *argv0;
...
+/*
+ * Copyright (C) 2016 Google Inc.
+ * Dan Cross <crossd@gmail.com>
+ * See LICENSE for license details.
+ */
+
+/*
+ * A tracking structure for growing lists of pointers.
+ */
+typedef struct PtrSlice PtrSlice;
```
Comment (dancrossnyc) [inline sys/include/libc.h]: I think we should call this 'PSlice'.

File: sys/include/libc.h [hunk 2]
```diff
@@ -758,3 +758,34 @@ extern char *argv0;
...
+
+struct PtrSlice {
+	void **ptrs;
+	size_t len;
+	size_t capacity;
+};
+
+void PtrSliceInit(PtrSlice *slice);
```
Comment (dancrossnyc) [inline sys/include/libc.h]: And this 'psliceinit' and etc.

File: sys/src/libc/port/slice.c [hunk 3]
```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (C) 2016 Google Inc.
+ * Dan Cross <crossd@gmail.com>
+ * See LICENSE for license details.
+ */
+
+#include <u.h>
+#include <libc.h>
+
+void PtrSliceInit(PtrSlice *slice)
```
Comment (dancrossnyc) [inline sys/src/libc/port/slice.c]: Can you clang-format this to Harvey style? Return value on separate line and all of that jazz...

File: sys/src/libc/port/slice.c [hunk 4]
```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (C) 2016 Google Inc.
+ * Dan Cross <crossd@gmail.com>
+ * See LICENSE for license details.
```
Comment (dancrossnyc) [inline sys/src/libc/port/slice.c]: I'd like to see this under ISC license. (I'll send you the text offline).

File: sys/include/libc.h [hunk 5]
```diff
@@ -758,3 +758,34 @@ extern char *argv0;
 /* this is used by sbrk and brk,  it's a really bad idea to redefine it */
 extern	char	end[];
 
+/* from BSD */
+void* reallocarray(void *base, size_t nel, size_t size);
+
+/* slice types and functions. */
```
Comment (dancrossnyc) [inline sys/include/libc.h]: Do we need the comment block in the header?

**Comment by rminnich** [comment]
Dan, I put the copyright in slice.c, not libc.h. I can move it to libc.h as you mentioned but ... that's unusual. See realloc.c for example, where it's copyright in code.

File: sys/include/libc.h (hunk 2 above)
Comment (rminnich) [inline sys/include/libc.h]: done

File: sys/include/libc.h (hunk 1 above)
Comment (rminnich) [inline sys/include/libc.h]: done

File: sys/src/libc/port/slice.c (hunk 3 above)
Comment (rminnich) [inline sys/src/libc/port/slice.c]: done

File: sys/src/libc/port/slice.c (hunk 4 above)
Comment (rminnich) [inline sys/src/libc/port/slice.c]: done, in slice.c

File: sys/include/libc.h (hunk 5 above)
Comment (rminnich) [inline sys/include/libc.h]: I removed it, I was not sure, but I think the function itself is fine.

**Comment by dancrossnyc** [comment]
On Sun, Aug 7, 2016 at 5:02 PM, ron minnich <notifications@github.com>
wrote:

> Dan, I put the copyright in slice.c, not libc.h. I can move it to libc.h
> as you mentioned but ... that's unusual. See realloc.c for example, where
> it's copyright in code
>
No, you are right; I had it reversed. Sorry about that!

**Comment by elbing** [comment]
Ok for me. Weird header, I guess it's GPL compatible. Since there's a fix over reallocarray, lgtm is reserved for both of you.
Thanks Dan.

**Comment by elbing** [comment]
LGTM

[Approved with PullApprove](https://pullapprove.com/Harvey-OS/harvey/pull-request/245/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Harvey-OS/harvey)

**Comment by dancrossnyc** [comment]
LG! Thanks, Ron.

**Pull request merged**